src/main/java/com/google/api/codegen/transformer/OutputTransformer.java

```diff
@@ -398,7 +398,23 @@ public void run() {
             valueSet.getId(),
             config.input());
       } else if (token == '{') {
-        throw new UnsupportedOperationException("map indexing not supported yet");
+        Preconditions.checkArgument(
```

Comment:
Is curly braces the syntax we want to have? I know it's (only slightly) easier for us to code here, but is it better for users?

In NodeJS, PHP, Python, and Go, (and I think C# and Ruby) we use square brackets to index both lists and maps. Is there a reason to break that convention here? 

@beccasaurus, what do you think?

Reply:
Incode samples already use curly braces to access map entries: https://github.com/googleapis/gapic-generator/blob/12e27eda15d2823684bba0a070252016bea9334e/src/test/java/com/google/api/codegen/testsrc/libraryproto/library_gapic.yaml#L522

Reply:
Sure, but the intent has always been to get rid of `samplecode_init_fields`, and in fact, it is going away with the microgenerators. I really don't think what we did here should have much, if any, weight on what we decide to do for samplegen.

The question, IMO, should be: what's most intuitive for our users? Given that it seems like the languages that we generate that support map subscripting use square brackets, I think there's a strong case for adhering to that widespread convention.

WDYT?

Reply:
I prefer curly braces. Happy to discuss further :)

First, yes, I agree we are gonna remove `sample_code_init_fields`, but my point was accessing a map field using `{}` is the pattern already adopted by gapic_config.yamls (In fact, wherever `InitCode` is used this pattern is adopted, so that includes smoke tests and the setup part for standalone samples too). I generally prefer not to break existing patterns unless we really need to.

Second, for explicitness. We have four ways to access nested elements all together: 1) access a field of a message, 2) access an entity of a resource name, 3) access a repeated field by giving an index, 4) access a map field by giving a key. The existing pattern has four different syntax for these four cases: `.` for 1), `%` for 2), `[]` for 3) and `{}` for 4) which is explicit and less error-prone. If we change 4) to use `[]`, same as 3), yes, it would be consistent with lots of real programming languages, but we will lose some explicitness.

Second point five (this is a bit of a personal philosophy so don't actually count as an argument), I think programming languages surfaces tend to lean towards being more general (as most of them use `[]` to access an array element and a map entry) to make themselves easier to use for developers, while configurations (yaml in our case) should lean towards being as explicit as possible to catch mistakes earlier.

Reply:
On point 1: if this affects other parts of the config after moving to annotations, then yes, that's an argument for `{ }`. But in the annotation world, will the brackets for maps continue to be used?

On points 2 and 2.5: I'm not sure that `{ }` is less error prone than `[ ]`: you could make the same errors in both when _writing_ the config, and it's only when running the generator that you get notified. What `{ }` arguably buys us is that when you _read_ the config, you're immediately reminded by the syntax that this is a map.

On point 2.5: I agree with your intent, but I'm wary of creating special-case syntax to replace common-use syntax in just our situation.

I'm still personally leaning towards `[ ]` simply because it's more in use in programming languages, though I could accept the alternative. My main point is that I want to make sure that whichever option we go with, we actually do so _consciously_, making an argument for it (and I say this as the person who was fine with `{ }` in the stub/TODO code).

@beccasaurus, as a user of this, do you have an opinion?

If not, let's not block on this, but try get additional opinions/feedback quickly so if we need to change it we can do so before it's in widespread use.

Reply:
> What `{ }` arguably buys us is that when you _read_ the config, you're immediately reminded by the syntax that this is a map.

This is what I like about `[ ]`

But I could go either way. No strong preference as `[ ]` is the most common syntax across languages for accessing both array items by index and map items by key.

Reply:
Hmmmm, it seems `[ ]` wins.
I guess we need to change to use `[ ]` everywhere within samples? Here's what I think we should do:
1. Not block on this PR
2. We reach consensus that only `[ ]` is accepted to access a map item by key everywhere within sample config. This includes `print`, `define`, `defaults` and `attributes`, in both gapic-generator and the micro generators.
3. Specifically for gapic-generator languages, we accept both `{}` and `[]` for `defaults` and `attributes` to not break exiting incode samples and smoke tests because these configurations share a same parser. 
4. Have a PR to implement 2 and 3 for gapic-generator languages
5. When we completely get rid of the smoke tests and the old way of configuring incode samples we will deprecate `{ }` altogether.
@vchudnov-g WDYT?

Reply:
Issue filed: #2600

Reply:
Cool, thank you!